suite/tests/client-interface/emulation_api_simple.dll.c

```diff
@@ -0,0 +1,165 @@
+ *
+ * * Neither the name of ARM Limited nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL ARM LIMITED OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ *
+ */
+
+/* A simple AArch64 client replacing 'dst = src0 & src1' with
+ * 'dst = !(!src0 | !src1)' to sanity test two core emulation API functions:
+ * - drmgr_insert_emulation_start()
+ * - drmgr_insert_emulation_end()
+ *
+ * Note this emulation client is for AArch64 only:
+ * XXX i#3173 Improve testing of emulation API functions
+ */
+
+#include "dr_api.h"
+#include "drmgr.h"
+#include <limits.h>
+#include <stdint.h>
+#include <string.h>
+#include <unistd.h>
+
+/* Events handlers */
+static void
+event_exit(void);
+
+static dr_emit_flags_t
+event_instruction_change(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
+                         bool translating);
+
+DR_EXPORT void
+dr_client_main(client_id_t id, int argc, const char *argv[])
+{
+    if (!drmgr_init())
+        DR_ASSERT(false);
+
+    dr_register_exit_event(event_exit);
+
+    if (!drmgr_register_bb_app2app_event(event_instruction_change, NULL))
+        DR_ASSERT(false);
+}
+
+static void
+event_exit(void)
+{
+    if (!drmgr_unregister_bb_app2app_event(event_instruction_change))
+        DR_ASSERT(false);
+    drmgr_exit();
+}
+
+static dr_emit_flags_t
+event_instruction_change(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
+                         bool translating)
+{
+    instr_t *instr, *next_instr;
+
+    for (instr = instrlist_first_app(bb); instr != NULL; instr = next_instr) {
+        /* We're deleting some instrs, so get the next first. */
+        next_instr = instr_get_next_app(instr);
+
+        /* We replace 'dst = src0 & src1' with 'dst = !(!src0 | !src1)'. We only
+         * transform AND instructions with unshifted, 64 bit register operands. */
+        if (instr_get_opcode(instr) != OP_and)
+            continue;
+        opnd_t dst = instr_get_dst(instr, 0);
+        opnd_t src0 = instr_get_src(instr, 0);
+        opnd_t src1 = instr_get_src(instr, 1);
+
+        if (!opnd_is_reg(src0) || instr_num_srcs(instr) != 4 ||
```

Comment:
maybe it would be good to explain this condition here a bit: below we only handle the AND with 4 source operands (2 regs, shift type and shift amount), with a shift amount of 0, and 64 bit registers. I don't think we need !opnd_is_reg(src0) here, as the first source operand for AND always is a reg.